File: UI/AvaCrmUi/src/app/services/note.service.ts
import { Injectable } from '@angular/core';
import {HttpClient, HttpParams} from '@angular/common/http';
import {Observable} from 'rxjs';
import {GlobalResponse, PaginatedResult, PaginationRequest} from '../models/base.model';
import {NoteCreateDto, NoteListDto, NoteUpdateDto} from '../dtos/CustomerManagment/note.models';
import {ApiAddressUtility} from '../utilities/api-address.utility';

@Injectable({
  providedIn: 'root'
})
export class NoteService {


  constructor(private http: HttpClient) { }

  // دریافت یادداشت بر اساس ID
  getById(id: number): Observable<GlobalResponse<NoteListDto>> {
    return this.http.get<GlobalResponse<NoteListDto>>(
      ApiAddressUtility.noteById(id)
    );
  }

  // دریافت یادداشت‌های یک مشتری
  getByCustomerId(
    customerId: number,
    request: PaginationRequest
  ): Observable<GlobalResponse<PaginatedResult<NoteListDto>>> {

    let params = new HttpParams()
      .set('PageNumber', request.pageNumber.toString())
      .set('PageSize', request.pageSize.toString());

    if (request.searchTerm) {
      params = params.set('SearchTerm', request.searchTerm);
    }
    if (request.sortColumn) {
      params = params.set('SortColumn', request.sortColumn);
    }
    if (request.sortDirection) {
      params = params.set('SortDirection', request.sortDirection);
    }

    return this.http.get<GlobalResponse<PaginatedResult<NoteListDto>>>(
      ApiAddressUtility.notesByCustomer(customerId),
      { params }
    );
  }

  // ایجاد یادداشت جدید
  create(createDto: NoteCreateDto): Observable<GlobalResponse<NoteListDto>> {
    return this.http.post<GlobalResponse<NoteListDto>>(
      ApiAddressUtility.createNote(),
      createDto
    );
  }

  // بروزرسانی یادداشت
  update(updateDto: NoteUpdateDto): Observable<GlobalResponse<NoteListDto>> {
    return this.http.put<GlobalResponse<NoteListDto>>(
      ApiAddressUtility.updateNote(),
      updateDto
    );
  }

  // حذف یادداشت
  delete(id: number): Observable<GlobalResponse<any>> {
    return this.http.delete<GlobalResponse<any>>(
      ApiAddressUtility.deleteNote(id)
    );
  }
}

File: AvaCrm.Application/Features/CustomerManagement/Customers/NoteService.cs
using AvaCrm.Application.DTOs.CustomerManagement.Notes;
using AvaCrm.Application.Pagination;
using AvaCrm.Application.Services;
using AvaCrm.Domain.Contracts.CustomerManagement;
using AvaCrm.Domain.Entities.CustomerManagement;
using Microsoft.EntityFrameworkCore;

namespace AvaCrm.Application.Features.CustomerManagement.Customers
{
    public class NoteService : BaseService, INoteService
	{
		private readonly INoteRepository _noteRepository;
		private readonly ICustomerRepository _customerRepository;

		public NoteService(
			IMapper mapper,
			INoteRepository noteRepository,
			ICustomerRepository customerRepository) : base(mapper)
		{
			_noteRepository = noteRepository;
			_customerRepository = customerRepository;
		}

		public async Task<GlobalResponse<NoteListDto>> GetById(long id, CancellationToken cancellationToken = default)
		{
			try
			{
				var note = await _noteRepository.GetById(id, cancellationToken);
				if (note == null)
					return new GlobalResponse<NoteListDto> { StatusCode = 404, Message = "یادداشت یافت نشد" };

				var result = _mapper.Map<NoteListDto>(note);
				return new GlobalResponse<NoteListDto> { StatusCode = 200, Data = result, Message = "عملیات موفق" };
			}
			catch (Exception ex)
			{
				return new GlobalResponse<NoteListDto> { StatusCode = 500, Message = $"خطا در دریافت اطلاعات: {ex.Message}" };
			}
		}

		public async Task<GlobalResponse<PaginatedResult<NoteListDto>>> GetByCustomerId(long customerId, PaginationRequest request, CancellationToken cancellationToken = default)
		{
			try
			{
				var query = _noteRepository.GetAll().Where(n => n.CustomerId == customerId);

				// Apply search filter
				if (!string.IsNullOrWhiteSpace(request.SearchTerm))
				{
					query = query.Where(n => n.Content.Contains(request.SearchTerm));
				}

				// Apply sorting
				query = request.SortDirection == "desc" ?
					query.OrderByDescending(n => n.CreationDate) :
					query.OrderBy(n => n.CreationDate);

				var totalCount = await query.CountAsync(cancellationToken);

				var notes = await query
					.Skip((request.PageNumber - 1) * request.PageSize)
					.Take(request.PageSize)
					.ToListAsync(cancellationToken);

				var result = new PaginatedResult<NoteListDto>
				{
					Items = _mapper.Map<List<NoteListDto>>(notes),
					TotalCount = totalCount,
					PageNumber = request.PageNumber,
					PageSize = request.PageSize
				};

				return new GlobalResponse<PaginatedResult<NoteListDto>>
				{
					StatusCode = 200,
					Data = result,
					Message = "عملیات موفق"
				};
			}
			catch (Exception ex)
			{
				return new GlobalResponse<PaginatedResult<NoteListDto>>
				{
					StatusCode = 500,
					Message = $"خطا در دریافت لیست یادداشت‌ها: {ex.Message}"
				};
			}
		}

		public async Task<GlobalResponse<NoteListDto>> Create(NoteCreateDto createDto, long userId, CancellationToken cancellationToken = default)
		{
			try
			{
				// Verify customer exists
				var customer = await _customerRepository.GetById(createDto.CustomerId, cancellationToken);
				if (customer == null)
					return new GlobalResponse<NoteListDto> { StatusCode = 404, Message = "مشتری یافت نشد" };

				var note = _mapper.Map<Note>(createDto);
				var created = await _noteRepository.Create(note, userId);
				var result = _mapper.Map<NoteListDto>(created);

				return new GlobalResponse<NoteListDto> { StatusCode = 201, Data = result, Message = "یادداشت با موفقیت ایجاد شد" };
			}
			catch (Exception ex)
			{
				return new GlobalResponse<NoteListDto> { StatusCode = 500, Message = $"خطا در ایجاد یادداشت: {ex.Message}" };
			}
		}

		public async Task<GlobalResponse<NoteListDto>> Update(NoteUpdateDto updateDto, long userId, CancellationToken cancellationToken = default)
		{
			try
			{
				var existing = await _noteRepository.GetById(updateDto.Id, cancellationToken);
				if (existing == null)
					return new GlobalResponse<NoteListDto> { StatusCode = 404, Message = "یادداشت یافت نشد" };

				_mapper.Map(updateDto, existing);
				await _noteRepository.Update(existing, userId, cancellationToken);
				var result = _mapper.Map<NoteListDto>(existing);

				return new GlobalResponse<NoteListDto> { StatusCode = 200, Data = result, Message = "یادداشت با موفقیت بروزرسانی شد" };
			}
			catch (Exception ex)
			{
				return new GlobalResponse<NoteListDto> { StatusCode = 500, Message = $"خطا در بروزرسانی یادداشت: {ex.Message}" };
			}
		}

		public async Task<GlobalResponse<ResponseResultGlobally>> Delete(long id, long userId, CancellationToken cancellationToken = default)
		{
			try
			{
				var note = await _noteRepository.GetById(id, cancellationToken);
				if (note == null)
					return new GlobalResponse<ResponseResultGlobally>
					{
						StatusCode = 404,
						Message = "یادداشت یافت نشد",
						Data = new ResponseResultGlobally { DoneSuccessfully = false }
					};

				await _noteRepository.Delete(id, userId);

				return new GlobalResponse<ResponseResultGlobally>
				{
					StatusCode = 200,
					Message = "یادداشت با موفقیت حذف شد",
					Data = new ResponseResultGlobally { DoneSuccessfully = true }
				};
			}
			catch (Exception ex)
			{
				return new GlobalResponse<ResponseResultGlobally>
				{
					StatusCode = 500,
					Message = $"خطا در حذف یادداشت: {ex.Message}",
					Data = new ResponseResultGlobally { DoneSuccessfully = false }
				};
			}
		}

	}
}